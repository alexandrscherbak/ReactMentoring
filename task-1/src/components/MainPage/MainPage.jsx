import React from 'react';
import FilmsTable from '../FilmsTable/FilmsTable.jsx';
import Footer from '../Footer/Footer.jsx';
import ResultsInfoPanel from '../ResultsInfoPanel/ResultsInfoPanel.jsx';
import SearchPanel from '../SearchPanel/SearchPanel.jsx';
import NoData from '../NoData/NoData.jsx';
import {connect} from 'react-redux';
import FilmsInfoActions from '../../redux/actions/films-info.action.js';
import { withRouter } from "react-router-dom";
import SortByActions from '../../redux/actions/sort-by.action.js';
import SearchValueActions from '../../redux/actions/search-value.action.js';
import SearchByActions from '../../redux/actions/search-by.action.js';

export class MainPage extends React.Component {
	constructor(props) {
		super(props);
	}

	componentDidMount() {
		const params = new URLSearchParams(this.props.location.search);
		const search = params.get('search');
		const searchBy = params.get('searchBy');
		const sortBy = params.get('sortBy');

		this.props.updateSearchValue(search);
		this.props.updateSearchBy(searchBy);
		this.props.updateSortBy(sortBy);

		this.props.updateFilmsInfo({
			search,
			searchBy,
			sortBy,
		});
	}

	selectMovie(filmId) {
		this.props.history.push(`/film/${filmId}`);
	}

	searchClicked() {
		const params = new URLSearchParams(this.props.location.search);
		const queryParams = [];
		queryParams.push(`search=${encodeURIComponent(this.props.searchValue)}`);
		queryParams.push(`searchBy=${encodeURIComponent(this.props.searchBy)}`);
		queryParams.push(`sortBy=${encodeURIComponent(this.props.sortBy)}`);
		this.props.history.push(`/search?${queryParams.join('&')}`);
		this.props.updateFilmsInfo({
			search: this.props.searchValue,
			searchBy: this.props.searchBy,
			sortBy: this.props.sortBy,
		});
	}

	render() {
		return (
			<React.Fragment>
				<SearchPanel
					searchClicked={() => this.searchClicked()}
				/>
				<ResultsInfoPanel
					count={this.props.filmsInfo.length}
				/>
				{
					this.props.filmsInfo.length ?
						<FilmsTable
							filmsInfo={this.props.filmsInfo}
							selectMovie={movie => this.selectMovie(movie.id)}
						/> :
						<NoData />
				}
				<Footer/>
			</React.Fragment>
		);
	}
}

export const mapStateToProps = state => {
	return {
		searchValue: state.searchValue,
		searchBy: state.searchBy,
		sortBy: state.sortBy,
		filmsInfo: state.filmsInfo,
	};
}

const mapDispatchToProps = dispatch => {
	return {
		updateFilmsInfo: requestParams => dispatch(FilmsInfoActions.loadFilmsInfo(requestParams)),
		updateSearchValue: searchValue => dispatch(SearchValueActions.setSearchValue(searchValue)),
		updateSearchBy: searchBy => dispatch(SearchByActions.setSearchBy(searchBy)),
		updateSortBy: sortBy => dispatch(SortByActions.setSortBy(sortBy)),
	};
}

export default withRouter(connect(
	mapStateToProps,
	mapDispatchToProps,
)(MainPage));
