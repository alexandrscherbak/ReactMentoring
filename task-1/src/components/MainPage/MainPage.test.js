import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { MainPage, mapStateToProps } from './MainPage.jsx';
import renderer from 'react-test-renderer';

jest.mock('../SearchPanel/SearchPanel.jsx', () => 'mock-search-panel');
jest.mock('../ResultsInfoPanel/ResultsSort/ResultsSort.jsx', () => () => 'ResultsSort');
Enzyme.configure({ adapter: new Adapter() });

const props = {
	filmsInfo: [
		{
			id: 2,
			poster_path: 'poster_path1',
			title: 'film1',
			release_date: '2018-04-01',
			genres: ['action', 'drama'],
		},
		{
			id: 3,
			poster_path: 'poster_path2',
			title: 'film2',
			release_date: '2018-04-01',
			genres: ['action', 'drama'],
		},
	],
	searchValue: 'searchValue',
	searchBy: 'searchBy',
	sortBy: 'sortBy',
	updateFilmsInfo: jest.fn(),
	updateSearchValue: jest.fn(),
	updateSearchBy: jest.fn(),
	updateSortBy: jest.fn(),
	match: {
		params: {
			filmId: 1,
		},
	},
	history: {
		push: jest.fn(),
	},
	location: {
		search: '?search=a%20b&searchBy=title&sortBy=vote_average',
	}
};

function setup() {
	return mount(<MainPage {...props}/>);
}

describe('MainPage', () => {
	test('snapshot', () => {
		const component = renderer.create(
			<MainPage {...props}/>,
		);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
	it('should call updateSearchValue in componentDidMount', () => {
		expect(props.updateSearchValue.mock.calls.length).toBe(1);
	});
	it('should call updateSearchBy in componentDidMount', () => {
		expect(props.updateSearchBy.mock.calls.length).toBe(1);
	});
	it('should call updateSortBy in componentDidMount', () => {
		expect(props.updateSortBy.mock.calls.length).toBe(1);
	});
	it('should call updateFilmsInfo in componentDidMount', () => {
		expect(props.updateFilmsInfo.mock.calls.length).toBe(1);
	});
	it('should call history.push if film clicked', () => {
		const enzymeWrapper = setup();
		const filmsTable = enzymeWrapper.find('FilmsTable');
		expect(props.history.push.mock.calls.length).toBe(0);
		filmsTable.props().selectMovie(props.filmsInfo[0]);
		expect(props.history.push.mock.calls.length).toBe(1);
	});
	it('should call updateFilmsInfo if search clicked', () => {
		const enzymeWrapper = setup();
		const searchPanel = enzymeWrapper.find('mock-search-panel');
		expect(props.updateFilmsInfo.mock.calls.length).toBe(3);
		searchPanel.props().searchClicked();
		expect(props.updateFilmsInfo.mock.calls.length).toBe(4);
	});
	it('should render FilmsTable if there are films', () => {
		const enzymeWrapper = setup();
		expect(enzymeWrapper.find('FilmsTable').length).toBe(1);
	});
	it('should render NoData if no films', () => {
		const enzymeWrapper = setup();
		enzymeWrapper.setProps({filmsInfo: []});
		expect(enzymeWrapper.find('NoData').length).toBe(1);
	});
	it('mapStateToProps should return props from state', () => {
		const state = {
			searchValue: 'searchValue',
			searchBy: 'searchBy',
			sortBy: 'sortBy',
			filmsInfo: 'filmsInfo',
		};
		expect(mapStateToProps(state)).toEqual({
			searchValue: 'searchValue',
			searchBy: 'searchBy',
			sortBy: 'sortBy',
			filmsInfo: 'filmsInfo',
		});
	});
});
