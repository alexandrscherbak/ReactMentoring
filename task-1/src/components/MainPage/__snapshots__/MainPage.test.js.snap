// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage snapshot 1`] = `
Array [
  <mock-search-panel
    searchClicked={[Function]}
  />,
  <div
    className="jsx-1938216586"
  >
    <div>
      <span>
        2
         movies found
      </span>
    </div>
    ResultsSort
  </div>,
  <div
    className="jsx-1758079044 films-table"
  >
    <div
      className="jsx-2546762946 film-info-container"
      onClick={[Function]}
    >
      <img
        alt="film1"
        className="jsx-2994430438"
        src="poster_path1"
      />
      <div
        className="jsx-2546762946 title-and-date"
      >
        <div
          className="jsx-433743616"
        >
          <span
            className="jsx-433743616"
          >
            film1
          </span>
        </div>
        <div
          className="jsx-617062702"
        >
          <span
            className="jsx-617062702"
          >
            2018
          </span>
        </div>
      </div>
      <div
        className="jsx-672986837"
      >
        <span
          className="jsx-672986837"
        >
          action
        </span>
      </div>
    </div>
    <div
      className="jsx-2546762946 film-info-container"
      onClick={[Function]}
    >
      <img
        alt="film2"
        className="jsx-2994430438"
        src="poster_path2"
      />
      <div
        className="jsx-2546762946 title-and-date"
      >
        <div
          className="jsx-433743616"
        >
          <span
            className="jsx-433743616"
          >
            film2
          </span>
        </div>
        <div
          className="jsx-617062702"
        >
          <span
            className="jsx-617062702"
          >
            2018
          </span>
        </div>
      </div>
      <div
        className="jsx-672986837"
      >
        <span
          className="jsx-672986837"
        >
          action
        </span>
      </div>
    </div>
  </div>,
  <div
    className="jsx-429413428"
  >
    <div
      className="jsx-2852891300"
    >
      <span
        className="jsx-2852891300"
      >
        netflixroulette
      </span>
    </div>
  </div>,
]
`;
